presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/FifoQueue.java

```diff
@@ -13,71 +13,120 @@
  */
 package com.facebook.presto.execution.resourceGroups;
 
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.Comparator;
 import java.util.Iterator;
-import java.util.LinkedHashSet;
+import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListSet;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterators.transform;
 
-// A queue with constant time contains(E) and remove(E)
+// A queue with constant time contains(E) and log time remove(E)
+@ThreadSafe
 final class FifoQueue<E>
         implements UpdateablePriorityQueue<E>
 {
-    private final Set<E> delegate = new LinkedHashSet<>();
+    private final Map<E, Entry<E>> map = new ConcurrentHashMap<>();
+    private final Set<Entry<E>> queue = new ConcurrentSkipListSet<>(Comparator.comparingLong(Entry::getSequence));
+
+    @GuardedBy("this")
+    private long sequence;
 
     @Override
-    public boolean addOrUpdate(E element, long priority)
+    public synchronized boolean addOrUpdate(E element, long priority)
     {
-        return delegate.add(element);
+        if (contains(element)) {
+            return false;
+        }
+        Entry<E> entry = new Entry<>(element, sequence);
+        sequence++;
+        map.put(element, entry);
+        checkState(queue.add(entry), "Current queue already contains the element");
+        return true;
     }
 
     @Override
     public boolean contains(E element)
     {
-        return delegate.contains(element);
+        return map.containsKey(element);
     }
 
     @Override
-    public boolean remove(E element)
+    public synchronized boolean remove(E element)
     {
-        return delegate.remove(element);
+        Entry entry = map.remove(element);
+        if (entry == null) {
+            return false;
+        }
+        checkState(queue.remove(entry), "Current queue does not contain the element");
+        return true;
     }
 
     @Override
-    public E poll()
+    public synchronized E poll()
     {
-        Iterator<E> iterator = delegate.iterator();
+        Iterator<Entry<E>> iterator = queue.iterator();
```

Comment:
Iterators of these concurrent data structures are weakly consistent, so according to the javadoc the traversal "may (but are not guaranteed to) reflect any modifications subsequent to construction". That is, if there are concurrent changes to these data structures we are not guaranteed to see them in the `poll` method.

Reply:
Yes, that's why I made those functions that can change the data structure synchronized. But of course, we need some "contract" to enforce that a single element should be immutable. Not sure if there is a way to do that.

Reply:
Making these methods `synchronized` will not help, because `synchronized` will grab the monitors of these queue instances while the concurrent data structures have their internal locks to synchronize access to their state. Since we don't have access to those internal locks we cannot change the the guarantees that they provide in their spec/javadoc.